### PR TITLE
Avoid further planning in case of result relation was pruned

### PR DESCRIPTION
**Pull request by Stolb27**
A result relation is a target relation for DML operations
in the appropriate queries. If the result relation is a member
of the join tree and is excluded by query predicates, such query
won't modify any tuples. So we could avoid spending resources
for planning remaining join tree relations, especially complex
sub queries. This gives a nice performance gain when planning
DML queries to tables with many leaf tables in the inheritance tree
(i.e. partitioned too). Because Postgres builds a separate plan for
each child relation first according to inheritance_planner function
logic.

There are several iterations to refactor this logic in the upstream
Postgres:
1. Plan Update/Delete query as SELECT query first to collect
list of affected children [[1]](https://www.postgresql.org/message-id/9d7c5112-cb99-6a47-d3be-cf1ee6862a1d@lab.ntt.co.jp).
3. Rework Modify Table node to handle tuples from any child
relation [[2]](https://www.postgresql.org/message-id/CA+HiwqHpHdqdDn48yCEhynnniahH78rwcrv1rEX65-fsZGBOLQ@mail.gmail.com)

So upstream Postgres and gpdb 7 is not interested in this patch.

### CODE REVIEW

**Comment by Stolb27** [comment]
Local tests have passed successfully for Postgres optimizer.

**Comment by Stolb27** [comment]
@InnerLife0, I've left test without memory limitation because manipulation of `gp_vmem_protect_limit` requires cluster restart. And test consumes 30 seconds only.

**Comment by InnerLife0** [comment]
This version of fix looks promising, but, as discussed, I don't like a code duplication. Though, I don't see much ways to refactor this.

File: src/test/regress/greenplum_schedule [hunk 1]
```diff
@@ -177,6 +177,7 @@ test: aggregate_with_groupingsets
 
 test: nested_case_null sort bb_mpph
 test: bb_memory_quota memconsumption
+test: DML_partition_pruning
```
Comment (seryozhasmirnov) [inline src/test/regress/greenplum_schedule]: All test names are in lower case. I think this test name should be in lower case too.

**Comment by andr-sokolov** [comment]
I have added debug output (https://github.com/andr-sokolov/gpdb/commit/68cbe511f2a4c58c0a6e9a7d0962e0fb051db8d7). The make_one_rel function is called once for each partition. After each call, memory consumption continuously increases. After the changes, it does not increase as fast as before the changes, but this means that the lack of memory will be on a different number of partitions. Maybe try to look for a solution in which memory consumption will not depend linearly on the number of partitions?
[before.log](https://github.com/arenadata/gpdb/files/8900483/before.log)
[after.log](https://github.com/arenadata/gpdb/files/8900482/after.log)

**Review by seryozhasmirnov — DISMISSED** [review]

File: src/test/regress/greenplum_schedule (hunk 1 above)
Comment (Stolb27) [reply src/test/regress/greenplum_schedule]: There is at least one opposite example: `src/test/regress/sql/DML_over_joins.sql`

**Comment by Stolb27** [comment]
@andr-sokolov, indeed, Postgres planner allocates memory mainly in the planner's Memory Context and rarely free it explicitly. There are cases when local buffers are freed and several optimizations for GECO-planning that uses separate memory context for memory-intensive parts of code. So I don't know how to exclude redundant memory consumption at all during planning. We tried to implement constraint exclusion logic closer to inheritance planner (see, #332 or [here](https://github.com/arenadata/gpdb/compare/adb-6.x...arenadata:c7870140a3a9ceaab0d36da0982f796d2e6d70eb?expand=1) that we discussed yesterday), but such solutions require to copy-paste even more source code. Though, such solutions have to consume less memory because makes decision earlier.

I've made a decision about placing code here, basing on memory profiling charts (that correlate with your logs, more charts in the tracker ticket) showing the main memory consumption during planning paths for sub-queries: 
memory_bytes

File: src/backend/optimizer/path/allpaths.c [hunk 2]
```diff
@@ -235,6 +247,23 @@ static void
 set_base_rel_sizes(PlannerInfo *root)
 {
 	Index		rti;
+	int			resultRelIndex = root->parse->resultRelation;
+	RelOptInfo *resultRel;
+
+	/*
+	 * If result relation is excluded by constraint exclusion we shouldn't
+	 * plan this sub-query
```
Comment (maksm90) [inline src/backend/optimizer/path/allpaths.c]: Please, add the notice about SELECT stmts that in this case resultRelIndex is zero, resultRel is NULL and all successive branch is omitted.

**Review by maksm90 — DISMISSED** [review]

File: src/backend/optimizer/path/allpaths.c (hunk 2 above)
Comment (Stolb27) [inline src/backend/optimizer/path/allpaths.c]:
```suggestion
	 * If result relation is excluded by constraint exclusion we shouldn't
	 * plan this sub-query. Result relation is set to zero for SELECT
	 * queries. simple_rel_array and simple_rte_array are filled by 
	 * zeroes during allocation and never use zero-indexed item.
	 * So, SELECT queries bypass this logic safely.
```

@maksm90, Is this enough description for this moment mentioned by yours?

Comment (maksm90) [reply src/backend/optimizer/path/allpaths.c]: Sure

**Review by maksm90 — DISMISSED** [review]

**Comment by Stolb27** [comment]
@seryozhasmirnov,
PR to upstream https://github.com/greenplum-db/gpdb/pull/13696 was merged to 6X_STABLE. I've updated this PR with changes, requested by upstream.

**Review by seryozhasmirnov — APPROVED** [review]

**Review by maksm90 — APPROVED** [review]

**Pull request merged**